tests/compiler/std/polyfills.optimized.wat

```diff
@@ -13,143 +12,55 @@
  (export "memory" (memory $0))
  (export "table" (table $0))
  (start $start)
- (func $~lib/polyfills/bswap16<u16> (; 1 ;) (; has Stack IR ;) (type $ii) (param $0 i32) (result i32)
-  (local $1 i32)
+ (func $~lib/polyfills/bswap<u8> (; 1 ;) (; has Stack IR ;) (type $FUNCSIG$i) (result i32)
+  (i32.const 170)
```

Comment:
I wonder why it leaves the function in place instead of pulling the result to the call site.

Reply:
bswap not inlined by binaryen even with so trivial body. I could add `@inline` but for 64-swap case this would pretty unnecessary for example

Reply:
Happy to take a look at a testcase where the binaryen inliner isn't doing what you want. Not sure what to look at in this PR though, can you direct me?

Reply:
@kripken I created simple fiddle for that:
https://webassembly.studio/?f=a2b2oz6y0al

Reply:
So with `-O3` (actually -O4 in binaryen) fiddle example produce this wasm:
```wast
(func $main/bswap (type $t2) (result i32)
    i32.const 16755455)
  (func $start (type $t1)
    call $main/bswap
    i32.const 65535
    i32.and
    i32.const 43775
    i32.ne
    if $I0
      i32.const 0
      i32.const 8
      i32.const 6
      i32.const 0
      call $~lib/env/abort
      unreachable
    end)
```

Reply:
So in this case `call $main/bswap` is redundant

Reply:
@MaxGraey I'm not sure how to use that fiddle. I click "build" and then look at the output wasm, and if I then run binaryen -O3 on that, it optimizes the start function into a nop (so it optimizes away everything). I also see the same if I remove `-O3` from the gulpfile.

How can I see the issue you are describing here?

Reply:
The strange thing is that it reduces the function, which originally has one parameter, to one that returns a constant, eliminating the parameter, but doesn't pull that constant to the call site then. I'd expect that it also goes the extra step to eliminate the function entirely.

Reply:
In the fiddle above, removing `-O3` from gulpfile.js emits the unoptimized module, which still has a parameter and doesn't return a constant.

```wat
  (func $main/bswap (type $t0) (param $p0 i32) (result i32)
    get_local $p0
    i32.const 8
    i32.shl
    get_local $p0
    i32.const 65535
    i32.and
    i32.const 8
    i32.shr_u
    i32.const 255
    i32.and
    i32.or
    i32.const 65535
    i32.and)
  (func $start (type $t2)
    i32.const 65450
    call $main/bswap
    ...
```

By adding `-O3` back to the gulpfile, Binaryen is invoked with `-O4` (AssemblyScript ups `-O3` to `-O4` automatically), which then exhibits the behaviour.

```wat
  (func $main/bswap (type $t2) (result i32)
    i32.const 43775)
  (func $start (type $t1)
    call $main/bswap
    ...
```

This somewhat seems like it's caused by the order passes are run.

Reply:
The main strange thing for me is why generated wasm from AS not reduced to this:
```ts
(module
  (type $t0 (func))
  (func $f0 (type $t0)
    nop)
  (table $table (export "table") 1 anyfunc)
  (memory $memory (export "memory") 1)
  (elem (i32.const 0) $f0)
  (data (i32.const 8) "\07\00\00\00m\00a\00i\00n\00.\00t\00s"))
```
In one pass? Any optimizations produce much bigger same result

Reply:
@kripken It will be useful if binaryen-js could return all used passes for current opt & shrink levels for make sure is all right and have right order. wdyt?

Reply:
@dcodeIO that's not the behavior of running Binaryen -O3 on the unoptimized wasm output, though - so I think something is going wrong with how AssemblyScript or the rest of the build system invokes Binaryen.

Is there a way for gulp and/or AssemblyScript to show the commands it issues, and the intermediate files? Specifically it would be good to see the wasm before and after each call to Binaryen's optimizer.

Reply:
Hmm, that's strange. The way compilation is performed is by first adding all the elements to Binaryen through the respective C-API calls, then setting optimize (=4) and shrink levels (=0) right before generating/printing the module. Nothing fancy happening there, like multiple calls or custom passes.

Need to look into this further I guess. Might be caused by manually adding the elements instead of parsing and re-emitting from a wat file or something. Or optimize / shrink levels are off somehow (fwiw: instead of using `-O3`, these can also be set directly through `--optimizeLevel=4` / `--shrinkLevel=0`, but this didn't make a difference when I tested).

Reply:
Seems adding `--noDebug` (this also calls `_BinaryenSetDebugInfo(false)`) works around the issue.

Reply:
Interesting... Binaryen has debug info off by default, so that should be a no-op, unless we either have a bug or you call it earlier with `true`?

Reply:
Yeah, it's enabled by default in AssemblyScript, unless `--noDebug` is specified. Probably should work the other way around then :)

Reply:
@kripken I saw you try improve `-O4` pass in binaryen. Could you include also this passes?
- licm (loop invariant code motion)

And may I ask you about this never used by default passes?
- reorder-functions (sorts functions by access frequency)
- ssa (ssa-ify variables so that they have a single assignment)

Does it make sense for us to add that passes for AssemblyScript?

Reply:
I'm still working on those other passes - need more testing and improvements. But I think I recall that

 * licm was not good for code size, and VMs do it, so it's not obvious the toolchain should.
 * reorder-functions has a comment in the source, that while it reduces binary size it increase gzip size. Needs more work there.

Reply:
Got it! Thanks for clarification